sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala

```diff
@@ -88,11 +88,29 @@ private case class MsSqlServerDialect() extends JdbcDialect with NoLegacyJDBCErr
         case e: Predicate => e.name() match {
           case "=" | "<>" | "<=>" | "<" | "<=" | ">" | ">=" =>
             val Array(l, r) = e.children().map {
-              case p: Predicate => s"CASE WHEN ${inputToSQL(p)} THEN 1 ELSE 0 END"
+              case p: Predicate => inputToCaseWhenSQL(p)
               case o => inputToSQL(o)
             }
             visitBinaryComparison(e.name(), l, r)
-          case "CASE_WHEN" => visitCaseWhen(expressionsToStringArray(e.children())) + " = 1"
+          case "CASE_WHEN" =>
+            // Since MsSqlServer cannot handle boolean expressions inside
+            // a CASE WHEN, it is necessary to convert those to an IIF
+            // expression that will return 1 or 0 depending on the result.
+            // Example:
+            // In:  ... CASE WHEN a = b THEN c = d ...
+            // Out: ... CASE WHEN a = b THEN IIF(c = d, 1, 0) ...
```

Comment:
```suggestion
            // Out: ... CASE WHEN a = b THEN IIF(c = d, 1, 0) ... = 1
```

Reply:
yep, agreed
I also forgot that I reverted back to CASE WHEN since I've put the util function for all connectors